### PR TITLE
16 Identity

### PR DESCRIPTION
**Pull request by jessywork**
User repository, service, controller, request model and response model.
LoginRequestModel
RegisterResponseModel
Exceptions

### CODE REVIEW

File: GrocifyApp.API/Controllers/AuthController.cs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        {
+            _configuration = configuration;
+
+            _userService = userService;
+        }
+
+        [HttpPost("register")]
+        public async Task<ActionResult<RegisterResponseModel>> Register([FromBody] UserRequestModel user)
+        {
+            var getUser = await _userService.GetUserByEmail(user.Email); //I cant use CheckEmailExists because this method needs id too
```
Comment (Catarina98) [inline GrocifyApp.API/Controllers/AuthController.cs]: remove comment

File: GrocifyApp.API/Models/ResponseModels/BadRequestModel.cs [hunk 2]
```diff
@@ -1,6 +1,6 @@
 ﻿namespace GrocifyApp.API.Models.ResponseModels
 {
-    public class BadRequestModel
+    public class BadRequestModel //this is request or response?
```
Comment (Catarina98) [inline GrocifyApp.API/Models/ResponseModels/BadRequestModel.cs]: response

File: GrocifyApp.API/Models/ResponseModels/UserResponseModel.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        public required Guid Id { get; set; }
+
+        /// <example>User Name</example>
+        public required string Name { get; set; }
+
+        /// <example>User Email</example>
+        public required string Email { get; set; }
+
+        /// <example>Password</example>
+        public required string Password { get; set; } = String.Empty; //string.Empty needed?
```
Comment (Catarina98) [inline GrocifyApp.API/Models/ResponseModels/UserResponseModel.cs]: only if you not make Password required

File: GrocifyApp.DAL/Exceptions/EmailExistsException.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+﻿namespace GrocifyApp.DAL.Exceptions
+{
+    public class EmailExistsException : Exception
+    {
+        public EmailExistsException()
+        {
+        }
+
+        public EmailExistsException(string message) : base(message)
```
Comment (Catarina98) [inline GrocifyApp.DAL/Exceptions/EmailExistsException.cs]: please remove both

File: GrocifyApp.DAL/Exceptions/EmailExistsException.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
+﻿namespace GrocifyApp.DAL.Exceptions
+{
+    public class EmailExistsException : Exception
+    {
+        public EmailExistsException()
```
Comment (Catarina98) [inline GrocifyApp.DAL/Exceptions/EmailExistsException.cs]: make it:

public EmailExistException(string message) : base(GenericConsts.Exceptions.EmailAlreadyTaken)
{
}

File: GrocifyApp.BLL/Implementations/UserService.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        public UserService(IUserRepository repository) : base(repository)
+        {
+            _uRepository = repository;
+        }
+
+        protected override async Task<bool> Validate(User user)
+        {
+            if (await _uRepository.CheckEmailExists(user.Email, user.Id))
+            {
+                throw new EmailExistsException(GenericConsts.Exceptions.EmailAlreadyTaken);
```
Comment (Catarina98) [inline GrocifyApp.BLL/Implementations/UserService.cs]: see EmailExistsException file comments, and remove this argument here

File: GrocifyApp.DAL/Repositories/Implementations/UserRepository.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+            {
+                user = await _dbContext.Users.SingleAsync(b => b.Email == email);
+            }
+
+            catch
+            {
+                user = null;
+            }
+
+            return user; //Possible null reference return
```
Comment (Catarina98) [inline GrocifyApp.DAL/Repositories/Implementations/UserRepository.cs]: make return object be nullable User?

you need to change UserService as well

**Review by Catarina98 — CHANGES_REQUESTED** [review]

File: GrocifyApp.API/Controllers/AuthController.cs [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        }
+
+        [HttpPost("login")]
+        public async Task<ActionResult<string>> Login([FromBody] LoginRequestModel user)
+        {
+            var getUser = await _userService.GetUserByEmail(user.Email);
+
+            if (getUser == null)
+            {
+                var errors = new List<string> { GenericConsts.Errors.UserPasswordIncorrect }; //and if we use the consts from DAL? If so, we only have one folder for consts, it seems more easier to search
```
Comment (jessywork) [inline GrocifyApp.API/Controllers/AuthController.cs]: Did you see this comment?

File: GrocifyApp.API/Models/RequestModels/UserRequestModel.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+        [RegularExpression(@"^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$", ErrorMessage = "Please enter a valid email address.")]
+        public required string Email { get; set; }
+
+        [StringLength(24, MinimumLength = 8)]
+        [RegularExpression(@"^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)[a-zA-Z\d]{8,}$", ErrorMessage = "Passwords must contains at least 8 characters and contain at least one uppercase letter and numbers.")]
+        public required string Password { get; set; }
+
+        [StringLength(24, MinimumLength = 8)]
+        [RegularExpression(@"^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)[a-zA-Z\d]{8,}$", ErrorMessage = "Passwords must contains at least 8 characters and contain at least one uppercase letter and numbers.")]
+        public string ConfirmPassword { get; set; } = String.Empty; //string.Empty needed? //i cant make this required because of line 84 of AuthController
```
Comment (jessywork) [inline GrocifyApp.API/Models/RequestModels/UserRequestModel.cs]: Did you see this comment?

File: GrocifyApp.API/Models/RequestModels/LoginRequestModel.cs [hunk 10]
```diff
@@ -0,0 +1,14 @@
+﻿using GrocifyApp.DAL.Models;
+using System.ComponentModel.DataAnnotations;
+
+namespace GrocifyApp.API.Models.RequestModels
+{
+    public class LoginRequestModel : BaseEntity
+    {
+        /// <example>joesmith@gmail.com</example>
+        [RegularExpression(@"^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$", ErrorMessage = "Please enter a valid email address.")] //const??
```
Comment (Catarina98) [inline GrocifyApp.API/Models/RequestModels/LoginRequestModel.cs]: const

File: GrocifyApp.API/Models/ResponseModels/BadResponseModel.cs [hunk 11]
```diff
@@ -1,6 +1,6 @@
 ﻿namespace GrocifyApp.API.Models.ResponseModels
 {
-    public class BadRequestModel
+    public class BadResponseModel //this is request or response?
```
Comment (Catarina98) [inline GrocifyApp.API/Models/ResponseModels/BadResponseModel.cs]: remove comment

File: GrocifyApp.API/Models/ResponseModels/UserResponseModel.cs [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+        /// <example>Id</example>
+        public required Guid Id { get; set; }
+
+        /// <example>User Name</example>
+        public required string Name { get; set; }
+
+        /// <example>User Email</example>
+        public required string Email { get; set; }
+
+        /// <example>Password</example>
```
Comment (Catarina98) [inline GrocifyApp.API/Models/ResponseModels/UserResponseModel.cs]: remove Password e ConfirmPassword

File: GrocifyApp.DAL/Exceptions/EmailExistsException.cs (hunk 4 above)
Comment (Catarina98) [reply GrocifyApp.DAL/Exceptions/EmailExistsException.cs]: try again

**Review by Catarina98 — CHANGES_REQUESTED** [review]

**Review by Catarina98 — APPROVED** [review]

**Pull request merged**